src/engagement_db_to_analysis/rapid_pro_advert_functions.py

```diff
@@ -0,0 +1,271 @@
+
+from core_data_modules.cleaners import Codes
+from core_data_modules.logging import Logger
+from core_data_modules.analysis import analysis_utils, AnalysisConfiguration as core_data_analysis_config
+
+from src.engagement_db_to_analysis.membership_group import (get_membership_groups_data)
+from src.engagement_db_to_analysis.cache import AnalysisCache
+
+from src.pipeline_configuration_spec import *
+
+
+log = Logger(__name__)
+
+CONSENT_WITHDRAWN_KEY = "consent_withdrawn"
+
+
+def _generate_weekly_advert_and_opt_out_uuids(participants_by_column, analysis_config,
+                                     google_cloud_credentials_file_path, membership_group_dir_path):
+    '''
+    Generates sets of weekly advert and  opt_out UUIDs to advertise to in rapid_pro.
```

Comment:
```suggestion
    Generates sets of weekly advert and  opt_out UUIDs to advertise to.
```
because this particular function isn't touching Rapid Pro.

Can you also add a brief sentence on exactly how they're generated? Something like "A participant is considered to have opted out if they are marked as 'consent_withdrawn' in the `participants_by_column` dataset. A participant is considered as being needed to advertise to if they are in the participants_by_column or a listening group and haven't opted out."

Reply:
Updated thanks.